File: dlib-sys/src/wrapper/shape_predictor.cpp
#pragma once

#include <memory>
#include <string>
#include <vector>

#include <dlib/image_processing/shape_predictor.h>

#include "cv_image.cpp"
#include "full_object_detection.cpp"
#include "rectangle.cpp"

namespace wrapper {

struct ShapePredictor {
  dlib::shape_predictor inner;

  ShapePredictor(const std::string file_path) {
    dlib::deserialize(file_path) >> inner;
  }

  FullObjectDetection function_call(const CvImage &cv_image,
                                    Rectangle rectangle) const {
    return FullObjectDetection(const_cast<dlib::shape_predictor &>(inner)(
        cv_image.inner, rectangle.inner));
  }

  // std::vector<Rectangle> functionCall(CvImage &cv_image) {
  //   std::vector<dlib::rectangle> dlib_rectangles = inner(cv_image.inner);
  //   std::vector<Rectangle> rectangles;
  //
  //   std::transform(dlib_rectangles.begin(), dlib_rectangles.end(),
  //                  std::back_inserter(rectangles),
  //                  [](dlib::rectangle x) { return Rectangle(x); });
  //
  //   return rectangles;
  // }
};
} // namespace wrapper

File: dlib-sys/src/shape_predictor.rs
use crate::{cv_image::CvImage, ffi, full_object_detection::FullObjectDetection, rectangle::Rectangle};
use autocxx::prelude::*;
use cxx::{let_cxx_string, CxxVector};
use std::{path::Path, pin::Pin};

pub struct ShapePredictor {
    pub(crate) inner: Pin<Box<crate::ffi::wrapper::ShapePredictor>>,
}

impl ShapePredictor {
    pub fn new<T: ffi::ToCppString>(file_path: T) -> Self {
        Self {
            inner: crate::ffi::wrapper::ShapePredictor::new(file_path).within_box(),
        }
    }

    /// Perform operator() call on c++ side hence this terrible name
    /// NOTE: Find a better name for this method
    pub fn function_call(&mut self, cv_image: &CvImage, rectangle: Rectangle) -> FullObjectDetection {
        FullObjectDetection {
            inner: self
                .inner
                .as_mut()
                .function_call(&cv_image.inner.as_ref(), rectangle.inner)
                .within_unique_ptr(),
        }
    }
}

unsafe impl Send for ShapePredictor {}
